---
title: "Multivariate analysis of genetic data: an introduction -- with Grant's Gazelle SNP Data"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

Convert BED/BIM/FAM to RAW
```{bash}
plink --bfile locality_grants --recodeA --out locality_grants
```

Find the current working directory with the following command. Our data should be in `(current directory)/grants_package/`

```{r}
getwd()
```

Follow http://adegenet.r-forge.r-project.org/files/PRstats/practical-MVAintro.1.0.pdf

Install the stable version of adegenet with dependencies

```{r}
install.packages("adegenet", dep=TRUE)
install.packages("ape", dep=TRUE)
install.packages("pegas", dep=TRUE)
install.packages("hierfstat", dep=TRUE)
install.packages("poppr", dep=TRUE)
install.packages("magrittr", dep=TRUE)
```

Load the useful packages 

```{r, echo=FALSE}
library("adegenet")
library("ape")
library("pegas")
library("hierfstat")
library("poppr")
library("magrittr")
```

Load PLINK .raw file

```{r}
data_prune <- read.PLINK("../grants_package/processed/species_prune.raw")
```


```{r}
pca_prune <- glPca(data_prune)
scatter(pca_prune, posi="topright")
tree <- nj(dist(as.matrix(data_prune)))
plot(root(tree, 58:60), cex=0.7)
```

Load STRUCTURE .stru file

```{r}
data_local_stru <- read.structure("../grants_package/processed/structure/locality_grants.stru", n.ind=95, n.loc=7566, onerowperind=FALSE, col.lab=1, col.pop=2, row.marknames=1, ask=FALSE)
popNames(data_local_stru)<-c("Amboseli", "Mkomazi_E", "Burigi", "Ikiri-Rungwa", "Aruba_Dam", "Ugalla", "Masai_Mara",	"Nairobi",	"Samburu",	"Tsavo",	"Sibiloi",	"Monduli", "Maswa","Mkomazi_W")

data_local_prune <- read.structure("../grants_package/processed/structure/locality_prune.stru", n.ind=95, n.loc=1933, onerowperind=FALSE, col.lab=1, col.pop=2, row.marknames=1, ask=FALSE)
popNames(data_local_prune)<-c("Amboseli", "Mkomazi_E", "Burigi", "Ikiri-Rungwa", "Aruba_Dam", "Ugalla", "Masai_Mara",	"Nairobi",	"Samburu",	"Tsavo",	"Sibiloi",	"Monduli", "Maswa","Mkomazi_W")


data_species_prune <- read.structure("../grants_package/processed/structure/species_prune.stru", n.ind=95, n.loc=1933, onerowperind=FALSE, col.lab=1, col.pop=2, row.marknames=1, ask=FALSE)
popNames(data_species_prune)<-c("g.granti", "g.robertsii", "notata", "petersii", "thomsons")

data_species_prune_no_og <- popsub(data_species_prune, blacklist = c("thomsons"))
```

## Asssesing population structure (page 15)

"From Population structure is traditionally measured and tested using F statistics, in particular the Fst, which measures population differentiation (as the proportion of allelic variance occuring between groups). The package hierfstat implements a wealth of F statistics and related tests, now designed to work natively with genind objects. The devel version of the package is required for these features." 


"We can now use different methods for assessing population structure. We first compute overall F statistics, and then use Goudet's G statistics to test the existence of population structure. Try to interpret the following statistics and graphics:"


```{r}
seppop(data_local_stru)
```


```{r}
pw_fst <- pairwise.fst(data_local_stru)
pw_fst_prune <- pairwise.fst(data_local_prune)

```


```{r}
pw_WCfst <- pairwise.WCfst(genind2hierfstat(data_local_prune))
pw_fst_CI <- boot.ppfst(genind2hierfstat(data_local_prune),nboot=100,quant=c(0.025,0.975),diploid=TRUE,dig=4)
```


http://popgen.nescent.org/DifferentiationSNP.html

### pairwise Fst
```{r}
genet.dist(data_local_prune, method = "WC84")
```

### Unsupervised clustering
using Kmeans and DAPC in adegenet 

```{r}
data_local_prune_no_og <- popsub(data_local_prune, blacklist = c("Ikiri-Rungwa"))
set.seed(20160308) # Setting a seed for a consistent result
grp <- find.clusters(data_local_prune_no_og, max.n.clust = 10, n.pca = 20, choose.n.clust = FALSE) 
grp$grp
dapc1 <- dapc(data_local_prune_no_og, grp$grp, n.pca = 20, n.da = 6) 
scatter(dapc1) # plot of the group
```


http://popgen.nescent.org/2015-05-18-Dist-SNP.html

### Section 2. Individual genetic distance: euclidean distance (dist {adegenet})

```{r}
distgenEUCL <- dist(data_local_prune, method = "euclidean", diag = FALSE, upper = FALSE, p = 2)
hist(distgenEUCL)
```



https://grunwaldlab.github.io/Population_Genetics_in_R/Minimum_Spanning_Networks.html

```{r}
local_amova <- poppr.amova(data_local_stru, data_local_stru$pop)

```



```{r}
data_local_prune_sub <- popsub(data_local_prune, blacklist = character(0))
data_local_prune_nomiss <- missingno(data_local_prune, type = 'mean')
data_local_prune_dist <- nei.dist(data_local_prune_nomiss, warning = TRUE)
min_span_net <- poppr.msn(data_local_prune_sub, data_local_prune_dist, showplot = FALSE, include.ties = TRUE)

set.seed(399)
plot_poppr_msn(data_local_prune,
               min_span_net,
               inds = "ALL",
               mlg = FALSE,
               gadj = 6,
               nodebase = 50,
               palette = rainbow,
               cutoff = NULL,
               quantiles = FALSE,
               beforecut = TRUE,
               layfun = igraph::layout_nicely)
```


http://adegenet.r-forge.r-project.org/files/PRstats/practical-MVAintro.1.0.pdf

p16

```{r}
data_local_prune.matFst <- pairwise.fst(data_local_prune, res.type = "matrix")
data_local_prune.tree <- nj(data_local_prune.matFst)
#plot(root(data_local_prune.tree, 58:60), type="unr")
plot(data_local_prune.tree, type="unr", tip.col=funky(nPop(data_local_prune)), font=2)
annot <- round(data_local_prune.tree$edge.length,2)
edgelabels(annot[annot>0], which(annot>0), frame="n")
add.scale.bar()


table.paint(data_local_prune.matFst, col.labels=1:14)


temp <- data_local_prune.matFst
diag(temp) <- NA
boxplot(temp, col=funky(nPop(data_local_prune)), las=3,
xlab="Population", ylab="Fst")

```


## sPCA

```{r}
coord <- read.csv("../grants_package/coord.csv")
data_local_prune$other$xy <- coord

spca_pw_dist <- spca(data_local_prune, type=5, d1=0, d2=5, scannf=FALSE)

barplot(spca_pw_dist$eig, main="A variant of the plotnn of sPCA eigenvalues", col=spectral(length(spca_pw_dist$eig)))
legend("topright", fill=spectral(2),
leg=c("Global structures", "Local structures"))
abline(h=0,col="grey")

plot(spca_pw_dist)
```

```{r}
library(akima)
x <- other(data_local_prune)$xy[,1]
y <- other(data_local_prune)$xy[,2]
temp <- interp(x, y, spca_pw_dist$li[,1], duplicate="mean")
image(temp, col=azur(100))
points(x,y)

myPal <- colorRampPalette(c("firebrick2", "white", "lightslateblue"))
annot <- function(){
title("sPCA - interpolated map of individual scores")
points(x,y)
}
filled.contour(temp, color.pal=myPal, nlev=50,
key.title=title("lagged nnscore 1"), plot.title=annot())
```


```{r}

coord_jittered <- data.frame(jitter(coord[,1]),jitter(coord[,2]))

data_local_prune_jit <- data_local_prune
data_local_prune_jit$other$xy <- coord_jittered



D <- dist(data_local_prune_jit$tab)

gab <- chooseCN(data_local_prune_jit$other$xy,ask=FALSE,type=2)
mon1 <- monmonier(data_local_prune_jit$other$xy,D,gab)

pco1 <- dudi.pco(D,scannf=FALSE,nf=1)
barplot(pco1$eig,main="Eigenvalues")

D <- dist(pco1$li)
mon1 <- monmonier(data_local_prune_jit$other$xy,D,gab)


```



```{r}
local.smry <- summary(data_local_prune)
plot(local.smry$Hobs, local.smry$Hexp, main="Observed vs expected heterozygosity")
abline(0,1,col="red")

local.X <- scaleGen(data_local_prune, NA.method="mean")
local.pca1 <- dudi.pca(local.X, cent=FALSE, scale=FALSE, scannf=FALSE, nf=2)
barplot(local.pca1$eig, main="local dataset - PCA eigenvalues",
col=heat.colors(length(local.pca1$eig)))

s.label(local.pca1$li)

plot.window(c(30,40),c(-8,1))
s.value(data_local_prune_jit$other$xy, local.pca1$li[,1], add.p=TRUE, cleg=0.5)

plot.window(c(30,40),c(-8,1))
s.value(data_local_prune_jit$other$xy, local.pca1$li[,2], add.p=TRUE, cleg=0.5)

local.graph <- chooseCN(data_local_prune_jit$other$xy,type=5,d1=0,d2=5, plot=FALSE, res="listw")

local.spca1 <- spca(data_local_prune_jit, cn=local.graph,scannf=FALSE, nfposi=2,nfnega=0)

plot.window(c(30,40),c(-8,1))
colorplot(data_local_prune_jit$other$xy, local.spca1$ls, axes=1:2, transp=TRUE, add=TRUE, cex=3)
title("sPCA - colorplot of PC 1 and 2nn(lagged scores)", col.main="yellow",line=-2, cex=2)
```


```{r}

data_species_prune <- read.structure("../grants_package/processed/structure/species_prune.stru", n.ind=95, n.loc=1933, onerowperind=FALSE, col.lab=1, col.pop=2, row.marknames=1, ask=FALSE)

coord <- read.csv("../grants_package/coord.csv")
coord <- data.frame(jitter(coord[,1]),jitter(coord[,2]))

data_species_prune$other$xy <- coord
data_species_prune_hyb <- popsub(data_species_prune, sublist = c("g.granti", "petersii"))

spca_pw_dist <- spca(data_species_prune_hyb, type=5, d1=0, d2=5, scannf=FALSE)

barplot(spca_pw_dist$eig, main="A variant of the plotnn of sPCA eigenvalues", col=spectral(length(spca_pw_dist$eig)))
legend("topright", fill=spectral(2),
leg=c("Global structures", "Local structures"))
abline(h=0,col="grey")

plot(spca_pw_dist)

library(akima)
x <- other(data_species_prune_hyb)$xy[,1]
y <- other(data_species_prune_hyb)$xy[,2]
temp <- interp(x, y, spca_pw_dist$li[,1], duplicate="mean")
image(temp, col=azur(100))
points(x,y)

myPal <- colorRampPalette(c("firebrick2", "white", "lightslateblue"))
annot <- function(){
title("sPCA - interpolated map of individual scores")
points(x,y)
}
filled.contour(temp, color.pal=myPal, nlev=50,
key.title=title("lagged nnscore 1"), plot.title=annot())

```

http://adegenet.r-forge.r-project.org/files/tutorial-dapc.pdf


```{r}
dapc3 <- dapc(data_species_prune, n.da=100, n.pca=20)
myCol <- rainbow(15)
par(mar=c(5.1,4.1,1.1,1.1), xpd=TRUE)
compoplot(dapc3, lab="", posi=list(x=12,y=-.01), cleg=.7)

temp <- which(apply(dapc3$posterior,1, function(e) all(e<0.5)))
lab <- pop(data_local_prune)
par(mar=c(8,4,5,1), xpd=TRUE)
compoplot(dapc3, subset=temp, cleg=.6, posi=list(x=0,y=1.2), lab=lab)


```



http://adegenet.r-forge.r-project.org/files/PRstats/practical-MVAgroups.1.0-allCmds.pdf


```{r}
grp1 <- find.clusters(data_species_prune_no_og, n.pca=40, n.clust=5)


dapc1 <- dapc(data_species_prune_no_og, pop=grp1$grp, scale=FALSE, n.pca=20, n.da=5)

scatter(dapc1, col=funky(6), scree.pca=TRUE)


tre1 <- nj(dist(tab(data_species_prune_no_og, freq=TRUE)))
plot(tre1, type="unr", show.tip=FALSE)
tiplabels(col=fac2col(grp1$grp), pch=20)

```